File: src/lib/notices/adapters/sofia.ts
import { Notice, NoticeProvider } from '../types';

export class SofiaAdapter implements NoticeProvider {
  name = 'SOFIA';
  
  private async getSessionCookie(): Promise<string> {
    try {
      const res = await fetch('https://sofia-briefing.aviation-civile.gouv.fr/sofia/pages/notamroute.html', {
        method: 'HEAD',
      });
      const setCookie = res.headers.get('set-cookie');
      if (!setCookie) return '';
      // Extract JSESSIONID
      const match = setCookie.match(/JSESSIONID=([^;]+)/);
      return match ? match[1] : '';
    } catch (e) {
      console.error('Error fetching SOFIA cookie:', e);
      return '';
    }
  }

  async getNotices(icao: string): Promise<Notice[]> {
    const now = new Date();
    const sessionId = await this.getSessionCookie();
    
    if (!sessionId) {
      console.error('Could not retrieve JSESSIONID from SOFIA');
      return [];
    }
    
    // Format dates
    const pad = (n: number) => n.toString().padStart(2, '0');
    // valid_from seems to require ISO-ish format: YYYY-MM-DDTHH:mm:ssZ
    const validFrom = now.toISOString().split('.')[0] + 'Z';
    
    // departure_date: dd-mm-yyyy
    const depDate = `${pad(now.getDate())}-${pad(now.getMonth() + 1)}-${now.getFullYear()}`;
    // departure_time: HHMM
    const depTime = `${pad(now.getHours())}${pad(now.getMinutes())}`;

    const params = new URLSearchParams();
    params.append(':operation', 'postAreaAeroPibRequest');
    params.append('isFromSofia', 'true');
    params.append('valid_from', validFrom);
    params.append('duration', '1200'); 
    params.append('traffic', 'VI');
    params.append('fl_lower', '0');
    params.append('fl_upper', '999');
    params.append('radius', '25');
    params.append('adep', icao);
    params.append('width', '15');
    params.append('aero[]', icao);
    params.append('operation', 'postAreaAeroPibRequest');
    params.append('target', '#aside-target');
    params.append('href', '/sofia/pages/notamroute.html');
    params.append('typeVol', 'L');
    params.append('departure_date', depDate);
    params.append('departure_time', depTime);
    params.append('lang', 'fr');
    params.append('routeVal', 'false');

    try {
      const res = await fetch('https://sofia-briefing.aviation-civile.gouv.fr/sofia', {
        method: 'POST',
        headers: {
          'Accept': 'application/json, text/javascript, */*; q=0.01',
          'Cookie': `JSESSIONID=${sessionId}`,
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'
        },
        body: params
      });

      if (!res.ok) {
          console.error('SOFIA Error', res.status, res.statusText);
          return [];
      }

      const json = await res.json();
      if (!json['status.message']) return [];

      let messageData;
      try {
          messageData = JSON.parse(json['status.message']);
      } catch (e) {
          console.error('Failed to parse inner status.message', e);
          return [];
      }

      const notices: Notice[] = [];
      const processedIds = new Set<string>();
      
      // Parsing logic
      const extractNotams = (obj: any) => {
          if (!obj) return;
          if (Array.isArray(obj)) {
              obj.forEach(item => extractNotams(item));
          } else if (typeof obj === 'object') {
              // Check if it's a NOTAM object
              if (obj.id && obj.series && obj.number && obj.itemE && !processedIds.has(obj.id)) {
                  processedIds.add(obj.id);
                  notices.push(this.mapToNotice(obj, icao));
              } 
              // Always dig deeper, even if we found a notam (though usually leaves are notams)
              // But notams don't contain other notams, so maybe else is fine.
              // However, the structure has arrays of objects.
              
              if (!obj.id) { // Optimize: dont search inside a NOTAM object
                Object.values(obj).forEach(val => extractNotams(val));
              }
          }
      };

      if (messageData.listnotams) {
          extractNotams(messageData.listnotams);
      }

      return notices.sort((a, b) => {
        // Sort by startValidity descending (newest first) or identifier
        return b.validFrom.localeCompare(a.validFrom);
      });

    } catch (error) {
      console.error('Error in SofiaAdapter:', error);
      return [];
    }
  }
  
  private mapToNotice(raw: any, defaultIcao: string): Notice {
      const series = raw.series || '';
      const number = raw.number ? raw.number.toString().padStart(4, '0') : '0000';
      const year = raw.year ? raw.year.toString().padStart(2, '0') : '00';
      const identifier = `${series}${number}/${year}`;
      
      const content = raw.multiLanguage?.itemE || raw.itemE || '';
      
      return {
          id: raw.id,
          icao: raw.itemA ? raw.itemA.split(' ')[0] : defaultIcao, 
          source: 'SOFIA',
          identifier,
          type: raw.type,
          validFrom: raw.startValidity,
          validTo: raw.endValidity,
          content: content,
          location: raw.coordinates,
          category: raw.qLine?.code23, 
      };
  }
}

File: crates/atc-book/src/adapters/sofia.rs
use reqwest::header::{CONTENT_TYPE, COOKIE, SET_COOKIE};
use serde_json::Value;

use crate::models::{Notice, NoticeSource};

/// Fetch NOTAMs from SOFIA for a given ICAO code.
pub async fn fetch_notices(icao: &str) -> Result<Vec<Notice>, String> {
    let client = reqwest::Client::builder()
        .build()
        .map_err(|e| format!("HTTP client error: {e}"))?;

    // Step 1: HEAD to get JSESSIONID
    let sofia_url = "https://sofia-briefing.aviation-civile.gouv.fr/sofia/pages/homepage.html";
    let resp = client
        .head(sofia_url)
        .send()
        .await
        .map_err(|e| format!("SOFIA HEAD failed: {e}"))?;

    let session_cookie = resp
        .headers()
        .get_all(SET_COOKIE)
        .iter()
        .filter_map(|v| v.to_str().ok())
        .find(|s| s.contains("JSESSIONID"))
        .map(|s| s.split(';').next().unwrap_or("").to_string())
        .ok_or_else(|| "JSESSIONID not found".to_string())?;

    // Step 2: POST PIB query
    let query_url = "https://sofia-briefing.aviation-civile.gouv.fr/sofia/api/notam";
    let body = format!(
        "operation=search&typeTraffic=VI&flLower=0&flUpper=999&radius=25&locations={}",
        icao.to_uppercase()
    );

    let resp = client
        .post(query_url)
        .header(COOKIE, &session_cookie)
        .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
        .body(body)
        .send()
        .await
        .map_err(|e| format!("SOFIA POST failed: {e}"))?;

    if !resp.status().is_success() {
        return Err(format!("SOFIA HTTP {}", resp.status()));
    }

    let json: Value = resp
        .json()
        .await
        .map_err(|e| format!("SOFIA JSON parse error: {e}"))?;

    // Recursively find NOTAM objects
    let mut notices = Vec::new();
    extract_notams(&json, icao, &mut notices);

    // Deduplicate by id
    notices.sort_by(|a, b| a.id.cmp(&b.id));
    notices.dedup_by(|a, b| a.id == b.id);

    // Sort by valid_from descending
    notices.sort_by(|a, b| b.valid_from.cmp(&a.valid_from));

    Ok(notices)
}

/// Recursively walk JSON to find NOTAM objects.
fn extract_notams(value: &Value, icao: &str, out: &mut Vec<Notice>) {
    match value {
        Value::Object(map) => {
            // Check if this looks like a NOTAM object (has "id" and "itemE")
            if let (Some(id_val), Some(item_e)) = (map.get("id"), map.get("itemE")) {
                let id = id_val.as_str().unwrap_or("").to_string();
                if !id.is_empty() {
                    let serie = map.get("serie").and_then(|v| v.as_str()).unwrap_or("");
                    let number = map.get("number").and_then(|v| v.as_str()).unwrap_or("");
                    let year = map.get("year").and_then(|v| v.as_str()).unwrap_or("");

                    let identifier = if !serie.is_empty() {
                        format!("{}{}/{}", serie, number, year)
                    } else {
                        id.clone()
                    };

                    let notice_type = map
                        .get("type")
                        .and_then(|v| v.as_str())
                        .unwrap_or("NOTAM")
                        .to_string();

                    let valid_from = map
                        .get("startValidity")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());

                    let valid_to = map
                        .get("endValidity")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());

                    let content = item_e.as_str().unwrap_or("").to_string();

                    let category = map
                        .get("qLine")
                        .and_then(|q| q.get("code23"))
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());

                    out.push(Notice {
                        id,
                        icao: icao.to_uppercase(),
                        source: NoticeSource::Sofia,
                        identifier,
                        notice_type,
                        valid_from,
                        valid_to,
                        content,
                        category,
                    });
                }
            }
            // Recurse into all values
            for v in map.values() {
                extract_notams(v, icao, out);
            }
        }
        Value::Array(arr) => {
            for v in arr {
                extract_notams(v, icao, out);
            }
        }
        _ => {}
    }
}
